src/input/json_to_storage.cc

```diff
@@ -336,21 +336,39 @@ StorageBase * JSONToStorage::make_storage(JSONPath &p, const Type::Record *recor
 {
     if (p.head()->type() == json_spirit::obj_type) {
     	const json_spirit::mObject & j_map = p.head()->get_obj();
-    	std::set<string> keys_to_processed;
+    	std::set<string> keys_to_process;
         json_spirit::mObject::const_iterator map_it;
         std::set<string>::iterator set_it;
 
         for( map_it = j_map.begin(); map_it != j_map.end(); ++map_it) {
-           keys_to_processed.insert(map_it->first);
+           keys_to_process.insert(map_it->first);
+        }
+
+        Type::Record::KeyIter key_it;
+        if ( record->has_key_iterator("TYPE", key_it) && record->auto_conversion_key_iter() != record->end() ) {
+            JSONPath type_path(p);
+            if (type_path.down( "TYPE" ) != NULL) {
+                try {
+                	if ( type_path.head()->get_str() != record->type_name() ) {
+                		xprintf(UsrErr, "Invalid value of TYPE key of record %s.", record->type_name().c_str());
```

Comment:
1) Don't use xprintf for error messages, use exceptions.
2) The condition is not necessary, I can imagine conversion to subrecord with
explicitly given TYPE:
	lower = IT::Record("lower","")
		.has_obligatory_type_key()
		.finish()
	IT::Record("Upper","")
		.has_obligatory_type_key()
		.declare_key("rec",lower, "")
		.finish()

	input="{TYPE=\"Lower\"}"

3) This automatic conversion is very tricky, we have to check it by unit tests. I have pushed one proposal which actually fails (not clear how to fix it, we can not provide default value to TYPE in has_obligatory_type_key,
possibly exclude TYPE key from check in allow_auto_conversion